tests/win_sec_attr.c

```diff
@@ -0,0 +1,98 @@
+/*
+    Copyright (c) 2015 Timothee "TTimo" Besset  All rights reserved.
+
+    Permission is hereby granted, free of charge, to any person obtaining a copy
+    of this software and associated documentation files (the "Software"),
+    to deal in the Software without restriction, including without limitation
+    the rights to use, copy, modify, merge, publish, distribute, sublicense,
+    and/or sell copies of the Software, and to permit persons to whom
+    the Software is furnished to do so, subject to the following conditions:
+
+    The above copyright notice and this permission notice shall be included
+    in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+
+
+*/
+
+#include "../src/nn.h"
+#include "../src/pair.h"
+#include "../src/pubsub.h"
+#include "../src/ipc.h"
+
+#include "testutil.h"
+
+#include <AccCtrl.h>
+#include <Aclapi.h>
+
+/*  Windows only. Custom SECURITY_ATTRIBUTES on a socket. */
+
+#define SOCKET_ADDRESS "ipc://win_sec_attr.ipc"
+
+int main ()
+{
+    int sb;
+    int sc;
+    SECURITY_ATTRIBUTES sec;
+    BOOL ret;
+    SID SIDAuthUsers;
+    DWORD SIDSize;
+    EXPLICIT_ACCESS xa;
+    PACL pACL;
+    DWORD ret2;
+    int ret3;
+
+    sc = test_socket (AF_SP, NN_PAIR);
+    test_connect (sc, SOCKET_ADDRESS);
+
+    sb = test_socket (AF_SP, NN_PAIR);
+
+    memset (&sec, 0, sizeof(sec));
+    sec.lpSecurityDescriptor = (PSECURITY_DESCRIPTOR)malloc (SECURITY_DESCRIPTOR_MIN_LENGTH);
+    ret = InitializeSecurityDescriptor (sec.lpSecurityDescriptor, SECURITY_DESCRIPTOR_REVISION);
+    nn_assert (ret);
+
+    SIDSize = sizeof (SIDAuthUsers);
+    ret = CreateWellKnownSid (WinAuthenticatedUserSid, NULL, &SIDAuthUsers, &SIDSize);
+    nn_assert (ret);
+
+    xa.grfAccessPermissions = GENERIC_READ | GENERIC_WRITE;
+    xa.grfAccessMode = SET_ACCESS;
+    xa.grfInheritance = SUB_CONTAINERS_AND_OBJECTS_INHERIT;
+    xa.Trustee.TrusteeForm = TRUSTEE_IS_SID;
+    xa.Trustee.TrusteeType = TRUSTEE_IS_WELL_KNOWN_GROUP;
+    xa.Trustee.ptstrName  = (LPSTR) &SIDAuthUsers;
+    ret2 = SetEntriesInAcl (1, &xa, NULL, &pACL);
+    nn_assert (ret2 == ERROR_SUCCESS);
```

Comment:
While this asserts that you can create this security attribute, I don't see how it validates that the IPC is actually created with that.  I'm not a windows guy, so I don't understand how you could even attempt to validate that -- how can you check the permissions of an Named Pipe?  I'd really like to see that verified though.